Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp

```diff
@@ -0,0 +1,419 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "MetalFunctionWriter.h"
+
+#include "AST.h"
+#include "ASTStringDumper.h"
+#include "ASTVisitor.h"
+#include <wtf/DataLog.h>
+#include <wtf/HashMap.h>
+#include <wtf/HashSet.h>
+#include <wtf/SetForScope.h>
+#include <wtf/text/StringBuilder.h>
+
+namespace WGSL {
+
+namespace Metal {
+
+constexpr bool dumpMetalCode = false;
+
+class FunctionDefinitionWriter : public AST::Visitor {
+public:
+    FunctionDefinitionWriter(StringBuilder& stringBuilder)
+        : m_stringBuilder(stringBuilder)
+    {
+    }
+
+    virtual ~FunctionDefinitionWriter() = default;
+
+    // Shader Module
+    void visit(AST::ShaderModule&) override;
+    // virtual void visit(GlobalDirective&);
+
+    // Attribute
+    void visit(AST::Attribute&) override;
+    // virtual void visit(AST::BindingAttribute&);
+    void visit(AST::BuiltinAttribute&) override;
+    // virtual void visit(AST::GroupAttribute&);
+    void visit(AST::LocationAttribute&) override;
+    void visit(AST::StageAttribute&) override;
+
+    // Declaration
+    // void visit(AST::Decl&) override;
+    void visit(AST::FunctionDecl&) override;
+    void visit(AST::StructDecl&) override;
+    void visit(AST::VariableDecl&) override;
+
+    // Expression
+    void visit(AST::Expression&) override;
+    void visit(AST::AbstractFloatLiteral&) override;
+    void visit(AST::AbstractIntLiteral&) override;
+    void visit(AST::ArrayAccess&) override;
+    // virtual void visit(AST::BoolLiteral&);
```

Comment:
We usually don't leave code commented out; we either uncomment it so it's used, or delete the line altogether. Is there a reason we're leaving it here?

Reply:
Fair enough, I just copied all the methods from the visitor, since we'll need to implement them later, but I'll just remove for now.